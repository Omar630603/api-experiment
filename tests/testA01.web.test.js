// test the ejs template for the index page

const request = require("supertest");
const app = require("../app");
const puppeteer = require("puppeteer");
const { toMatchImageSnapshot } = require("jest-image-snapshot");
expect.extend({ toMatchImageSnapshot });
const options = {
  showPrefix: false,
  showMatcherMessage: true,
  showStack: false,
};
require("dotenv").config();

let browser;
let page;

beforeAll(async () => {
  browser = await puppeteer.launch({
    headless: true,
    slowMo: 0,
    devtools: false,
  });
  page = await browser.newPage();
  await page.setDefaultTimeout(10000);
  await page.setDefaultNavigationTimeout(20000);
});

afterAll(async () => {
  await browser.close();
});

describe("Testing the index page", () => {
  it("should have the right title", async () => {
    const response = await request(app).get("/");
    expect(response.text).toContain("API-Experiment | Home");
  });

  it("should receive a message and display it", async () => {
    const response = await request(app).get("/?message=Hello test");
    expect(response.text).toContain("Hello test");
  });
});

describe("Testing the index page with puppeteer", () => {
  it("should have the right title", async () => {
    await page.goto(`http://localhost:${process.env.PORT}/`);
    const title = await page.title();
    expect(title).toBe("API-Experiment | Home");
  });

  it("should receive a message and display it", async () => {
    await page.goto(`http://localhost:${process.env.PORT}/?message=Hello test`);
    const message = await page.$eval(".message", (el) => el.textContent);
    expect(message).toBe("Hello test");
  });

  it("matches the expected styling", async () => {
    await page.goto(`http://localhost:${process.env.PORT}/`);
    const screenshot = await page.screenshot({ fullPage: true });
    expect(
      screenshot,
      `The web styling for the index page is not correct check the file "/tests/images/__image_snapshots_A01__/__diff_output__/index-page-styling-diff.png" to find the difference`,
      options
    ).toMatchImageSnapshot({
      customDiffConfig: { threshold: 0.1 },
      customSnapshotsDir: "tests/images/__image_snapshots_A01__",
      customSnapshotIdentifier: "index-page-styling",
    });
  });

  it("should have nav bar with 2 links", async () => {
    await page.goto(`http://localhost:${process.env.PORT}/`);
    const navBar = await page.$eval("nav", (el) => el.textContent);
    expect(navBar).toContain("Home");
    expect(navBar).toContain("Products");
  });

  it("should have a button to the products page", async () => {
    await page.goto(`http://localhost:${process.env.PORT}/`);
    await page.click(".btn.btn-primary");
    const url = await page.url();
    expect(url).toBe(`http://localhost:${process.env.PORT}/products`);
  });
});
