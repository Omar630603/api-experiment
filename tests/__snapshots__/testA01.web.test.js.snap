// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the index page with puppeteer match index-page snapshot: index-page 1`] = `
"<!DOCTYPE html><html lang="en"><head>
    <meta charset="UTF-8">
    <title>API-Experiment | Home</title>
    <link href="https://fonts.googleapis.com/css?family=Poppins&amp;display=swap" rel="stylesheet">
    <link rel="stylesheet" href="/styles/main.css" type="text/css">
  </head>
  <body>
    <nav>
      <ul>
        <li><a href="/">Home</a></li>
        <li><a href="/products">Products</a></li>
      </ul>
    </nav>
    <main><div class="container">
  <h1 class="title">Welcome to my beautiful web page!</h1>
  
  <div class="alert alert-warning">
    <p class="message">
      Type a value for the property message next to this page URL
      <br>
      /?message=Hello
    </p>
  </div>
  
</div>

<div class="container flex-col-container">
  <h2 class="subtitle">I can show products from my database in MongoDB</h2>
  <p class="description">Click on this button below to go the products page</p>
  <a href="/products" class="btn btn-primary">Products</a>
</div>
</main>
  

</body></html>"
`;

exports[`Testing the index page with puppeteer matches the products button snapshot: products-button 1`] = `"<a href="/products" class="btn btn-primary">Products</a>"`;
